src/throttle/amoc_throttle_process.erl

```diff
@@ -143,46 +100,38 @@ handle_continue(maybe_run_fn, State) ->
     NewState = maybe_run_fn(State),
     {noreply, NewState, timeout(NewState)}.
 
+-spec terminate(term(), state()) -> ok.
+terminate(_, State) -> %% Flush all pending actions
+    maybe_run_fn(State#state{can_run_fn = true, max_n = infinity}),
+    ok.
+
 -spec format_status(gen_server:format_status()) -> gen_server:format_status().
-format_status(#{state := #state{} = State} = FormatStatus) ->
-    ScheduleLen = length(State#state.schedule),
-    ScheduleRevLen = length(State#state.schedule_reversed),
-    State1 = setelement(#state.schedule, State, ScheduleLen),
-    State2 = setelement(#state.schedule_reversed, State1, ScheduleRevLen),
-    FormatStatus#{state := State2}.
+format_status(#{state := State} = FormatStatus) ->
+    FormatStatus#{state := printable_state(State)}.
 
 %%------------------------------------------------------------------------------
 %% internal functions
 %%------------------------------------------------------------------------------
 
-initial_state(Name, Interval, Rate) when Rate > 0 ->
-    NewRate = case Rate < 5 of
-                  true ->
-                      Msg = <<"too low rate, please reduce NoOfProcesses">>,
-                      internal_error(Msg, Name, Rate, Interval),
-                      Rate;
-                  false ->
-                      Rate
-              end,
-    Delay = case {Interval, Interval div NewRate, Interval rem NewRate} of
-                {0, _, _} -> 0; %% limit only No of simultaneous executions
-                {_, I, _} when I < 10 ->
-                    Message = <<"too high rate, please increase NoOfProcesses">>,
-                    internal_error(Message, Name, Rate, Interval),
-                    10;
-                {_, DelayBetweenExecutions, 0} -> DelayBetweenExecutions;
-                {_, DelayBetweenExecutions, _} -> DelayBetweenExecutions + 1
-            end,
-    #state{interval = Interval, n = NewRate, max_n = NewRate, delay_between_executions = Delay}.
-
-merge_state(#state{interval = I, delay_between_executions = D, n = N, max_n = MaxN},
-            #state{n = OldN, max_n = OldMaxN} = OldState) ->
+%% - If both `MaxN' and `Delay' are numbers, this will be the actual rate/interval.
+%%
+%% - If `MaxN' is a number and `Delay' is infinity, we mean to pause the process,
+%%      see how at `maybe_start_timer/1' a delay of infinity will set `can_run_fn = false'.
+%%
+%% - If `MaxN' is infinity and `Delay' is zero, we mean no limits to throttling,
```

Comment:
we don't care if `Delay` is zero, we check only if it's not equal to `infinity`:
- If `MaxN' is infinity and `Delay' is a number